test/hotspot/jtreg/runtime/os/TestExplicitPageAllocation.java

```diff
@@ -0,0 +1,173 @@
+                if (pageSize.equals(tracePageSize)) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    public static int readPageCount(String filename, String pageSize) throws Exception {
+        BufferedReader d = new BufferedReader(new FileReader(filename));
+        int pageCount = Integer.parseInt(d.readLine());
+        if (pageCount == 0) {
+            System.out.println("No pages configured for " + pageSize + "kB");
+        }
+        return pageCount;
+    }
+
+    public static void doSetup() throws Exception {
+        // Large page sizes
+        File[] directories = new File(DIR_HUGE_PAGES).listFiles(File::isDirectory);
+        for (File dir : directories) {
+            String pageSizeFileName = dir.getName();
+            Matcher matcher = HUGEPAGE_PATTERN.matcher(pageSizeFileName);
+            matcher.find();
+            String pageSize = matcher.group(1);
+            assert pageSize != null;
+            int availablePages = readPageCount(DIR_HUGE_PAGES + pageSizeFileName + "/free_hugepages", pageSize);
+            int index = Long.numberOfTrailingZeros(Long.parseLong(pageSize) * 1024);
+            pageSizes[index] = true;
+            pageCount[index] = availablePages;
+        }
+        // OS vm page size
+        vmPageSizeIndex = Long.numberOfTrailingZeros(wb.getVMPageSize());
+        pageSizes[vmPageSizeIndex] = true;
+        pageCount[vmPageSizeIndex] = Integer.MAX_VALUE;
+    }
+
+    public static void testCase(int index) throws Exception {
+        String pageSize = sizeFromIndex(index);
+        ProcessBuilder pb = ProcessTools.createJavaProcessBuilder("-Xlog:pagesize",
+                                                                  "-XX:LargePageSizeInBytes=" + pageSize,
+                                                                  "-XX:+UseParallelGC",
+                                                                  "-XX:+UseLargePages",
+                                                                  "-Xmx2g",
+                                                                  "-Xms1g",
+                                                                  "-version");
+        OutputAnalyzer output = new OutputAnalyzer(pb.start());
+        output.shouldHaveExitValue(0);
+        for (int i = index; i >= vmPageSizeIndex; i--) {
+            if (pageSizes[i]) {
+                String size = sizeFromIndex(i);
+                System.out.println("Checking allocation for " + size);
+                if (!checkOutput(output, size)) {
+                    if (requiredPageCount(i) > pageCount[i]) {
+                       continue;
+                    }
+                    errorMessage += "TestCase Failed for " + size + " page allocation, ";
+                } else {
+                    System.out.println("TestCase Passed for pagesize: " + pageSize + ", allocated pagesize: " + size);
+                    break;
+                }
```

Comment:
The test is still failing for 16G pages: 
```
TestCase Passed for pagesize: 16G, allocated pagesize: 64K
----------System.err:(12/796)----------
java.lang.AssertionError: TestCase Failed for 16G page allocation, 
	at runtime.os.TestExplicitPageAllocation.testCase(TestExplicitPageAllocation.java:170)
	at runtime.os.TestExplicitPageAllocation.main(TestExplicitPageAllocation.java:69)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	at java.base/java.lang.reflect.Method.invoke(Method.java:577)
	at com.sun.javatest.regtest.agent.MainWrapper$MainThread.run(MainWrapper.java:127)
	at java.base/java.lang.Thread.run(Thread.java:828)
```

The problem is that `requiredPageCount(i)` will return 0 for 16G pages because the heap is just 1G, so the test will fill in the `errorMessage` and then the test will continue. The test not breking after setting the `errorMessage` also looks a bit wrong. At least it made it a bit harder to spot the error.